app/wasm/test/messages_test.go

```diff
@@ -336,3 +336,170 @@ func TestSwap(t *testing.T) {
 	}
 
 }
+
+func TestSwapMultiHop(t *testing.T) {
+	actor := RandomAccountAddress()
+	osmosis, ctx := SetupCustomApp(t, actor)
+	epsilon := 1e-3
+
+	fundAccount(t, ctx, osmosis, actor, defaultFunds)
+
+	poolFunds := []sdk.Coin{
+		sdk.NewInt64Coin("uosmo", 12_000_000),
+		sdk.NewInt64Coin("ustar", 240_000_000),
+	}
+	// 20 star to 1 osmo
+	starPool := preparePool(t, ctx, osmosis, actor, poolFunds)
+
+	// 2 osmo to 1 atom
+	poolFunds2 := []sdk.Coin{
+		sdk.NewInt64Coin("uatom", 6_000_000),
+		sdk.NewInt64Coin("uosmo", 12_000_000),
+	}
+	atomPool := preparePool(t, ctx, osmosis, actor, poolFunds2)
+
+	// Multi-hop
+	// Estimate 1st swap rate
+	uosmo := poolFunds[0].Amount.ToDec().MustFloat64()
+	ustar := poolFunds[1].Amount.ToDec().MustFloat64()
+	swapRate1 := ustar / uosmo
```

Comment:
This is not really how to estimate. It only works for spot prices (as you found out).

The way you can estimate/calculate this this:

```
expectedOut1 := uosmo - uosmo * ustar / (ustar + amountIn.Input.ToDec().MustFloat64())
expectedOut2 := uatom2 - uosmo2 * uatom2 / (uosmo2 + expectedOut)
```